client/tx/tx_test.go

```diff
@@ -121,49 +123,110 @@ func TestBuildUnsignedTx(t *testing.T) {
-		WithFees("50stake").
-		WithMemo("memo").
-		WithChainID("test-chain")
+	// create a new key using a mnemonic generator and test if we can reuse seed to recreate that account
+	_, seed, err := kr.NewMnemonic(from1, keyring.English, path, hd.Secp256k1)
+	requireT.NoError(err)
+	requireT.NoError(kr.Delete(from1))
+	info1, _, err := kr.NewMnemonic(from1, keyring.English, path, hd.Secp256k1)
+	requireT.NoError(err)
 
-	msg := banktypes.NewMsgSend(info.GetAddress(), sdk.AccAddress("to"), nil)
-	txn, err := tx.BuildUnsignedTx(txf, msg)
-	require.NoError(t, err)
+	info2, err := kr.NewAccount(from2, seed, "", path, hd.Secp256k1)
+	requireT.NoError(err)
 
-	t.Log("should failed if txf without keyring")
-	err = tx.Sign(txf, from, txn)
-	require.Error(t, err)
+	pubKey1 := info1.GetPubKey()
+	pubKey2 := info2.GetPubKey()
+	requireT.NotEqual(pubKey1.Bytes(), pubKey2.Bytes())
+	t.Log("Pub keys:", pubKey1, pubKey2)
 
-	txf = tx.Factory{}.
-		WithKeybase(kr).
+	txfNoKeybase := tx.Factory{}.
 		WithTxConfig(NewTestTxConfig()).
 		WithAccountNumber(50).
 		WithSequence(23).
 		WithFees("50stake").
 		WithMemo("memo").
 		WithChainID("test-chain")
+	txfDirect := txfNoKeybase.
+		WithKeybase(kr).
+		WithSignMode(signingtypes.SignMode_SIGN_MODE_DIRECT)
+	txfAmino := txfDirect.
+		WithSignMode(signingtypes.SignMode_SIGN_MODE_LEGACY_AMINO_JSON)
+	msg1 := banktypes.NewMsgSend(info1.GetAddress(), sdk.AccAddress("to"), nil)
+	msg2 := banktypes.NewMsgSend(info2.GetAddress(), sdk.AccAddress("to"), nil)
+	txb, err := tx.BuildUnsignedTx(txfNoKeybase, msg1, msg2)
+	requireT.NoError(err)
+	txbSimple, err := tx.BuildUnsignedTx(txfNoKeybase, msg2)
+	requireT.NoError(err)
 
-	t.Log("should succeed if txf with keyring")
-	err = tx.Sign(txf, from, txn)
-	require.NoError(t, err)
+	testCases := []struct {
+		name         string
+		txf          tx.Factory
+		txb          client.TxBuilder
+		from         string
+		overwrite    bool
+		expectedPKs  []cryptotypes.PubKey
+		matchingSigs []int // if not nil, check matching signature against old ones.
+	}{
+		{"should fail if txf without keyring",
+			txfNoKeybase, txb, from1, true, nil, nil},
+		{"should fail for non existing key",
+			txfAmino, txb, "unknown", true, nil, nil},
+		{"should succeed with keyring Amino",
+			txfAmino, txbSimple, from1, true, []cryptotypes.PubKey{pubKey1}, nil},
+
+		/**** test double sign AMINO ****/
+		{"should sign tx2 Amino",
+			txfAmino, txb, from1, true, []cryptotypes.PubKey{pubKey1}, nil},
+		{"should append a second signature and not overwrite Amino",
+			txfAmino, txb, from2, false, []cryptotypes.PubKey{pubKey1, pubKey2}, []int{0, 0}},
+		{"should overwrite a signature Amino",
+			txfAmino, txb, from2, true, []cryptotypes.PubKey{pubKey2}, []int{1, 0}},
+		{"should fail to append a signature with different mode Direct",
+			txfDirect, txb, from1, false, []cryptotypes.PubKey{}, nil},
+
+		/**** test double sign DIRECT
+		  signing transaction with more than 2 signers should fail in DIRECT mode ****/
```

Comment:
we could maybe remove these `/**/` comments and just use better testCase.name.

Reply:
test name with many words is not very readable. I still want to describe the context. Other `/*...*/` comments are used to create sections. I went through the test names, and they look good IMHO, I reordered them. Please point to a test name which is not clear enough and I will think about better names.

Reply:
Admittedly a few like `"should sign tx2 Amino"` and `"should fail to sign tx2 DIRECT"` don't sound so very clear to me e.g. why is it `tx2` and not just `tx`?

One small thing: you use different caps for DIRECT (somehwere appears as `Direct`, elsewhere as `DIRECT`), how about picking one and sticking to it? Plus, maybe you could use the sign mode as a kind of prefix in the test name e.g. `DIRECT: fails to sign tx`, etc

Reply:
Yes, good idea. I will use `Direct`.